src/hotspot/share/gc/serial/serialHeap.cpp

```diff
@@ -838,27 +704,83 @@ void SerialHeap::collect(GCCause::Cause cause) {
 }
 
 void SerialHeap::do_full_collection(bool clear_all_soft_refs) {
-   do_full_collection(clear_all_soft_refs, OldGen);
+  if (GCLocker::check_active_before_gc()) {
+    return;
+  }
+  do_full_collection_no_gc_locker(clear_all_soft_refs);
 }
 
-void SerialHeap::do_full_collection(bool clear_all_soft_refs,
-                                    GenerationType last_generation) {
-  do_collection(true,                   // full
-                clear_all_soft_refs,    // clear_all_soft_refs
-                0,                      // size
-                false,                  // is_tlab
-                last_generation);       // last_generation
-  // Hack XXX FIX ME !!!
-  // A scavenge may not have been attempted, or may have
-  // been attempted and failed, because the old gen was too full
-  if (gc_cause() == GCCause::_gc_locker && incremental_collection_failed()) {
-    log_debug(gc, jni)("GC locker: Trying a full collection because scavenge failed");
-    // This time allow the old gen to be collected as well
-    do_collection(true,                // full
-                  clear_all_soft_refs, // clear_all_soft_refs
-                  0,                   // size
-                  false,               // is_tlab
-                  OldGen);             // last_generation
+void SerialHeap::do_full_collection_no_gc_locker(bool clear_all_soft_refs) {
+  IsSTWGCActiveMark gc_active_mark;
```

Comment:
`IsSTWGCActiveMark active_gc_mark;`in`do_young_collection_no_gc_locker`, just choose one and be consistent with it